pinot-core/src/main/java/org/apache/pinot/core/operator/filter/AndFilterOperator.java

```diff
@@ -60,12 +63,22 @@ protected BlockDocIdSet getFalses() {
     List<BlockDocIdSet> blockDocIdSets = new ArrayList<>(_filterOperators.size());
     for (BaseFilterOperator filterOperator : _filterOperators) {
       if (filterOperator.isResultEmpty()) {
-        blockDocIdSets.add(new MatchAllDocIdSet(_numDocs));
-      } else {
-        blockDocIdSets.add(filterOperator.getFalses());
+        return new MatchAllDocIdSet(_numDocs);
       }
+      if (filterOperator.isResultMatchingAll()) {
+        continue;
+      }
+      if (_nullHandlingEnabled) {
+        blockDocIdSets.add(
+            new OrDocIdSet(Arrays.asList(filterOperator.getTrues(), filterOperator.getNulls()), _numDocs));
```

Comment:
In a lot of cases, `getNulls()` can return `EmptyDocIdSet`, in which case we should fall back to the case without null handling